### PR TITLE
Provide placeholder generics for traits in "no method found for type parameter" suggestions

### LINKED ISSUE #132407: "no method found for type parameter" missing generic args
### Code

```rust
trait Trait<I> {
    fn method(&self) {}
}

fn foo<T>(value: T) {
    value.method()
}
```

### Current output

```rust
error[E0599]: no method named `method` found for type parameter `T` in the current scope
 --> src/lib.rs:6:11
  |
5 | fn foo<T>(value: T) {
  |        - method `method` not found for this type parameter
6 |     value.method()
  |           ^^^^^^ method not found in `T`
  |
  = help: items from traits can only be used if the type parameter is bounded by the trait
help: the following trait defines an item `method`, perhaps you need to restrict type parameter `T` with it:
  |
5 | fn foo<T: Trait>(value: T) {
  |         +++++++
```

### Desired output

```rust
error[E0599]: no method named `method` found for type parameter `T` in the current scope
 --> src/lib.rs:6:11
  |
5 | fn foo<T>(value: T) {
  |        - method `method` not found for this type parameter
6 |     value.method()
  |           ^^^^^^ method not found in `T`
  |
  = help: items from traits can only be used if the type parameter is bounded by the trait
help: the following trait defines an item `method`, perhaps you need to restrict type parameter `T` with it:
  |
5 | fn foo<T: Trait<_>>(value: T) {
  |         ++++++++++
```

### PR DESCRIPTION
**Pull request by dianne**
In the diagnostics for the error ``no method named `method` found for type parameter `T` in the current scope [E0599]``, the compiler will suggest adding bounds on `T` for traits that define a method named `method`. However, these suggestions didn't include any generic arguments, so applying them would result in a `missing generics for trait` or `missing lifetime specifier` error. This PR adds placeholder arguments to the suggestion in such cases. Specifically, I tried to base the placeholders off of what's done in suggestions for when generics are missing or too few are provided:
- The placeholder for a parameter without a default is the name of the parameter.
- Placeholders are not provided for parameters with defaults.

Placeholder arguments are enclosed in `/*` and `*/`, and the applicability of the suggestion is downgraded to `Applicability::HasPlaceholders` if any placeholders are provided.

Fixes #132407

### CODE REVIEW

File: compiler/rustc_hir_typeck/src/method/suggest.rs [hunk 1]
```diff
@@ -3872,22 +3872,50 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                                 param.name.ident(),
                             ));
                             let bounds_span = hir_generics.bounds_span_for_suggestions(def_id);
+                            let mut applicability = Applicability::MaybeIncorrect;
+                            // Format the path of each suggested candidate, providing placeholders
+                            // for any generic arguments without defaults.
+                            let candidate_strs: Vec<String> = candidates
+                                .iter()
+                                .map(|cand| {
+                                    use ty::GenericParamDefKind::{Const, Lifetime, Type};
```
Comment (fmease) [inline compiler/rustc_hir_typeck/src/method/suggest.rs:3881]: You should be able to simplify this by using `identity_for_item` over `generics_of`. Compare:

https://github.com/rust-lang/rust/blob/ef972a346668ed4234d1a43ed4ad7ca4e9c58d51/compiler/rustc_hir_analysis/src/hir_ty_lowering/errors.rs#L253-L263

**Review by fmease — CHANGES_REQUESTED** [review]
Could you surround the placeholder arguments with `/*` and `*/` to highlight the fact that they are placeholders? E.g., `Trait2</* '_, A, B */>`. Otherwise users may take them literally and wonder whether the suggestion is wrong if it leads to more errors. It's not a hypothetical btw, people have reported such suggestions as bugs.

This would also be consistent with the diagnostic suggestion for E0220 (associated type not found). See https://github.com/rust-lang/rust/pull/126575/commits/02a2f0272736c65b51996b0c6e7cbd991ac21d5c were I impl'ed that.

File: compiler/rustc_hir_typeck/src/method/suggest.rs (hunk 1 above)
Comment (fmease) [reply compiler/rustc_hir_typeck/src/method/suggest.rs:3881]: Ah, disregard, that wouldn't consider generic parameter defaults.

**Comment by dianne** [comment]
The placeholders should now be surrounded by `/*` and `*/`. I'll open a PR to do that for E0107 too, if that's alright. That's what I was using as a reference, and it looks easy to fix.

**Review by fmease — APPROVED** [review]
Thanks! Sorry for the slight delay. I was just dumbfounded that we don't have an API for that already somewhere. I guess one could use `GenericArgs::identity_for_item` plus `Generics::own_args_no_defaults` but er, that would perform much worse in this case.

I have small stylistic suggestions but nothing major. Lastly, could you squash your commits into one?

**Comment by dianne** [comment]
Done. Thanks!
@rustbot ready

**Pull request merged**